services/meta/client.go

```diff
@@ -539,7 +539,23 @@ func (c *Client) Authenticate(username, password string) (*UserInfo, error) {
 	userInfo := c.cacheData.User(username)
 	c.mu.RUnlock()
 	if userInfo == nil {
-		return nil, ErrUserNotFound
+		// Instead of instantly returning when the username is not found, compare against a valid bcrypt hash
+		// so that we don't leak whether the username was valid via how quickly the server responds.
+		// See https://cwe.mitre.org/data/definitions/208.html for more details.
+
+		// Create fake bcrypt hash with Version=2, Cost=<current cost setting>, Salt+Hash=<doesn't really matter>.
+		fakeHash := fmt.Sprintf("$2$%02d$00000000000000000000000000000000000000000000000000000", bcryptCost)
+
+		// Compare the password against the fake bcrypt hash so we eat up the right time according to bcryptCost.
+		err := bcrypt.CompareHashAndPassword([]byte(fakeHash), []byte(password))
```

Comment:
In benchmarking this implementation, I observed that ignoring the error seemed to instruct the compiler to discard the entire call to `bcrypt.CompareHashAndPassword`.